### PR TITLE
sql: backfill partial indexes

### LINKED ISSUE #51852: sql: backfill partial indexes

### PR DESCRIPTION
**Pull request by mgartner**
#### sql: backfill partial indexes

This commit adds support for backfilling partial indexes. Adding partial
indexes to tables that already have rows will now correctly backfill the
partial index such that it only indexes rows that satisfy the partial
index.

In order to backfill partial indexes, the backfiller first generates a
typed expression from the partial index predicate string. Then, for each
row in the table, the typed predicate expression is evaluated. If the
result is true, index entries are generated for the partial index.

Fixes #51852 

Release note: None

#### sql: add tests for computed columns that reference other new columns

This commit adds tests for adding computed columns that reference other
new columns added in the same `ALTER TABLE` statement. These cases were
previously untested.

Release note: None

### CODE REVIEW

File: pkg/sql/schemaexpr/expr.go [hunk 1]
```diff
@@ -131,3 +135,84 @@ func ExtractColumnIDs(
 
 	return colIDs, err
 }
+
+// NameResolver is used to replace unresolved names in expressions with
+// IndexedVars.
+type NameResolver struct {
+	evalCtx    *tree.EvalContext
+	tableID    sqlbase.ID
+	cols       []sqlbase.ColumnDescriptor
```
Comment (rohany) [inline pkg/sql/schemaexpr/expr.go:144]: We want to generally carry slices of `*ColumnDescriptor`, since these are getting expensive to copy around

File: pkg/sql/logictest/testdata/logic_test/partial_index [hunk 2]
```diff
@@ -455,3 +455,41 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
 100  110  100  foo  true
 300  330  300  foo  true
 400  400  400  foo  true
+
+#### Backfill a partial index.
+
+statement ok
+CREATE TABLE e (a INT, b INT)
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/partial_index:462]: unfortunately, we have two codepaths of how this all runs. One like this, and then one that kicks in when doing schema changes on a new table created in the same transaction. You should pretty much duplicate your tests with those cases.

File: pkg/sql/backfill/backfill.go [hunk 3]
```diff
@@ -468,6 +528,36 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 			return nil, nil, err
 		}
 
+		iv.CurSourceRow = ib.rowVals
+
+		// For the current row, make a list of the indexes to encode entries
+		// for.
+		indexesToEncode := ib.added
+		if len(ib.predicates) > 0 {
+			indexesToEncode = make([]sqlbase.IndexDescriptor, 0, len(ib.added))
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:537]: This should be allocated and stored on the index builder so that we don't allocate it per row

File: pkg/sql/backfill/backfill.go [hunk 4]
```diff
@@ -366,19 +379,65 @@ func (ib *IndexBackfiller) Init(
...
+				id := ib.cols[i].ID
 				if idx.ContainsColumnID(id) ||
 					idx.GetEncodingType(desc.PrimaryIndex.ID) == sqlbase.PrimaryIndexEncoding {
 					valNeededForCol.Add(i)
 				}
 			}
+
+			// If the index is a partial index, convert and store the predicate
+			// as a TypedExpr.
+			if idx.IsPartial() {
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:392]: Can you take a look at that changes made to the ColumnBackfiller in https://github.com/cockroachdb/cockroach/pull/51865 -- that is the end state for what i imagine these backfillers will look like when there are user defined types in these expressions. Right now, this wouldn't work with user defined types. I'm also OK with it not handling it right now and then opening an issue for it to handle later

**Review by rohany — CHANGES_REQUESTED** [review]

File: pkg/sql/backfill/backfill.go (hunk 4 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go:392]: Nice, looks like that should be pretty straightforward to add. I've added a note in [the issue I'm using to track partial indexes + UDTs](https://github.com/cockroachdb/cockroach/issues/51804). I'll probably go through in one swoop and add a bunch of UDT+partial index tests and fix whatever is broken, if that's ok with you. There's a number of existing bugs at the moment.

Comment (rohany) [reply pkg/sql/backfill/backfill.go:392]: Yup, seems reasonable to me.

File: pkg/sql/schemaexpr/expr.go (hunk 1 above)
Comment (mgartner) [reply pkg/sql/schemaexpr/expr.go]: Done.

File: pkg/sql/logictest/testdata/logic_test/partial_index (hunk 2 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:462]: Ahh good to know. Done.

File: pkg/sql/backfill/backfill.go [hunk 5]
```diff
@@ -360,6 +359,15 @@ type IndexBackfiller struct {
...
 	evalCtx *tree.EvalContext
+	cols    []sqlbase.ColumnDescriptor
+
+	// predicates is a map of indexes to partial index predicate expressions. It
+	// includes entries for partial indexes only.
+	predicates map[sqlbase.IndexID]tree.TypedExpr
+	// indexesToEncode is a list of indexes to encode entries for a given row.
+	// It is a field of IndexBackfiller to avoid allocating a slice for each row
+	// backfilled.
+	indexesToEncode []sqlbase.IndexDescriptor
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:370]: this maybe should be a slice of pointers -- copying descriptors here seems somewhat expensive (since we're going to do it per row)

File: pkg/sql/backfill/backfill.go [hunk 6]
```diff
@@ -372,36 +380,135 @@ func (ib *IndexBackfiller) ContainsInvertedIndex() bool {
...
-	numCols := len(desc.Columns)
-	cols := desc.Columns
+	// Initialize ib.cols and ib.colIdxMap.
+	ib.initCols(desc)
+
+	// Initialize ib.added.
+	valNeededForCol := ib.initIndexes(desc)
+
+	// Convert any partial index predicate strings into expressions.
+	predicates, predicateRefColIDs, err := schemaexpr.MakePartialIndexExprs(ctx, ib.added, ib.cols, desc, evalCtx, semaCtx)
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:399]: nit: long line

File: pkg/sql/backfill/backfill.go [hunk 7]
```diff
@@ -482,6 +613,36 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
...
+		// If there are any partial indexes being added, make a list of the
+		// indexes that the current row should be added to.
+		if len(ib.predicates) > 0 {
+			ib.indexesToEncode = ib.indexesToEncode[:0]
+			for i := range ib.added {
+				idx := &ib.added[i]
+				if !idx.IsPartial() {
+					// If the index is not a partial index, all rows should have
+					// an entry.
+					ib.indexesToEncode = append(ib.indexesToEncode, *idx)
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:627]: yeah we shouldn't copy this

File: pkg/sql/backfill/backfill.go [hunk 8]
```diff
@@ -491,7 +652,7 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 		if buffer, err = sqlbase.EncodeSecondaryIndexes(
 			ib.evalCtx.Codec,
 			tableDesc.TableDesc(),
-			ib.added,
+			ib.indexesToEncode,
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:655]: oh my goodness its because this takes a slice of nonpointers. That's unfortunate. I'm not sure whats the best thing to do here. I feel like these copies are going to show up in a profile for sure.

File: pkg/sql/logictest/testdata/logic_test/computed [hunk 9]
```diff
@@ -700,6 +700,17 @@ SELECT * FROM x ORDER BY a
 3 6 7
 6 12 10
 
+# Multiple ADD COLUMNs can only reference columns that are defined previously.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/computed:703]: Maybe pull these tests into a separate commit

File: pkg/sql/logictest/testdata/logic_test/partial_index [hunk 10]
```diff
@@ -4,6 +4,10 @@
 statement ok
 SET experimental_partial_indexes=on
 
+# TODO(mgartner): remove this once enums are fully supported.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/partial_index:7]: This TODO is unnecessary

File: pkg/sql/logictest/testdata/logic_test/partial_index [hunk 11]
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
...
 300  330  300  foo  true
 400  400  400  foo  true
+
+#### Backfill a partial index.
+
+statement ok
+CREATE TABLE e (a INT, b INT)
+
+statement ok
+INSERT INTO e VALUES (1, 10);
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/partial_index:469]: just do 1 insert statement with all of the values?

File: pkg/sql/logictest/testdata/logic_test/partial_index [hunk 12]
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
...
+CREATE INDEX a_b_gt_30_idx ON e (a) WHERE b > 30
+
+query II
+SELECT * FROM e@a_b_gt_30_idx WHERE b > 30
+----
+4  40
+5  50
+6  60
+
+#### Backfill a partial index when a new table is created in the same
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/partial_index:486]: nit: not sure we have a precedent for using comments like this

File: pkg/sql/logictest/testdata/logic_test/partial_index [hunk 13]
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
...
+INSERT INTO e VALUES (3, 30);
+INSERT INTO e VALUES (4, 40);
+INSERT INTO e VALUES (5, 50);
+INSERT INTO e VALUES (6, 60);
+
+statement ok
+CREATE INDEX a_b_gt_30_idx ON e (a) WHERE b > 30
+
+query II
+SELECT * FROM e@a_b_gt_30_idx WHERE b > 30
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/partial_index:480]: Is there some way that we can scan the index manually to ensure that there aren't extra KV's? Maybe something like the tests in `schema_changer_test.go`

**Review by rohany — CHANGES_REQUESTED** [review]
looking good! mostly small requests here. @lucy-zhang might also want to take a peek at this.

File: pkg/sql/logictest/testdata/logic_test/partial_index (hunk 10 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:7]: Why's that?

File: pkg/sql/backfill/backfill.go (hunk 5 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go:370]: It has to be a `[]sqlbase.IndexDescriptor` unless we change the type signature of `sqlbase.EncodeSecondaryIndexes`. I can go ahead and do that, but it'll be a pretty big change.

File: pkg/sql/backfill/backfill.go (hunk 6 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go]: Done.

File: pkg/sql/backfill/backfill.go (hunk 5 above)
Comment (rohany) [reply pkg/sql/backfill/backfill.go:370]: Let's measure it and see (see my comment below). If it does show up as expensive, then I think adding an another function that takes all of the indexes, and then a slice of index ordinals to write would be clean and not affect the API too much.

File: pkg/sql/backfill/backfill.go (hunk 8 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go:662]: I'm going to try to change this to take a slice of pointers. Maybe it won't be so bad...

File: pkg/sql/logictest/testdata/logic_test/computed (hunk 9 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/computed:703]: Done.

File: pkg/sql/logictest/testdata/logic_test/partial_index (hunk 12 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:483]: I don't know where I got this from... Fixed.

File: pkg/sql/logictest/testdata/logic_test/partial_index (hunk 11 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:466]: Done.

File: pkg/sql/backfill/backfill.go (hunk 8 above)
Comment (rohany) [reply pkg/sql/backfill/backfill.go:662]: I think it might be bad :D We use this mainly when we have a table descriptor, so we take that slice of indexes right off of there.

File: pkg/sql/backfill/backfill.go (hunk 5 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go:370]: I tested **(A)** regular index backfill on 100k rows on master vs. **(B)** a partial index backfill on 100k rows where every row is in the partial index.

#### A

master

#### B

partial_index_backfill

File: pkg/sql/logictest/testdata/logic_test/partial_index (hunk 13 above)
Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:475]: Forcing the index should be the same thing, AFAIK. This test will fail if a row is added to the partial index when it shouldn't be.

Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/partial_index:478]: I've added a comment to make it more clear that this will be a full partial index scan.

File: pkg/sql/backfill/backfill.go (hunk 5 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go:370]: I've created an issue to track reducing these allocations in the future: https://github.com/cockroachdb/cockroach/issues/52425

File: pkg/sql/backfill/backfill.go [hunk 14]
```diff
@@ -372,36 +380,142 @@ func (ib *IndexBackfiller) ContainsInvertedIndex() bool {
...
+
+	// Initialize ib.added.
+	valNeededForCol := ib.initIndexes(desc)
+
+	evalCtx := flowCtx.NewEvalCtx()
+	var predicates map[sqlbase.IndexID]tree.TypedExpr
+	var predicateRefColIDs sqlbase.TableColSet
+
+	// Install type metadata in the target descriptors, as well as resolve any
+	// user defined types in the column expressions.
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:439]: [nit] any user defined types in partial index expressions

File: pkg/sql/backfill/backfill.go [hunk 15]
```diff
@@ -410,20 +524,38 @@ func (ib *IndexBackfiller) Init(
...
+	// reset in BuildIndexEntriesChunk for every row added.
+	ib.indexesToEncode = ib.added
+	if len(ib.predicates) > 0 {
+		ib.indexesToEncode = make([]sqlbase.IndexDescriptor, 0, len(ib.added))
 	}
-	ib.colIdxMap = make(map[sqlbase.ColumnID]int, len(cols))
-	for i := range cols {
-		ib.colIdxMap[cols[i].ID] = i
+
+	ib.types = make([]*types.T, len(ib.cols))
```
Comment (rohany) [inline pkg/sql/backfill/backfill.go:549]: Is this used anywhere?

Comment (mgartner) [reply pkg/sql/backfill/backfill.go:549]: Yes here: https://github.com/cockroachdb/cockroach/blob/3b193feb06feac8d4e908a409b1abfad11d57699/pkg/sql/backfill/backfill.go#L481

It's already there on master.

File: pkg/sql/backfill/backfill.go (hunk 14 above)
Comment (mgartner) [reply pkg/sql/backfill/backfill.go]: Good catch. Done.

**Review by rohany — APPROVED** [review]

**Review by thoszhang — COMMENTED** [review]
The backfiller changes LGTM.

> Reviewed 2 of 16 files at r1, 2 of 4 files at r3, 18 of 18 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52213#-:-ME1YSfQEyM5QnVCMUAL:btldzc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang, @mgartner, and @rohany)

**Comment by mgartner** [comment]
Welp just before I thought I passed this level, I think I encountered the final boss, a flaky test failure of the test where an index with a UDT is backfilled within a transaction.

It looks like its a problem with computed columns with UDTs as well on master, so I submitted a ticket here: https://github.com/cockroachdb/cockroach/issues/52539

I want to get this PR in, but I don't want to merge a flaky test.

@rohany any ideas what could be going on? Is `flowCtx.TypeResolverFactory.Descriptors.ReleaseAll(ctx)` not releasing the leases correctly?

**Comment by rohany** [comment]
I don't want your PR here to get hung up. Can we just comment out the test here that would flake with a TODO pointing to the new issue?

**Comment by mgartner** [comment]
Yes, done.
